File: docs/VCG/src/vcg_client.cpp
// src/vcg_client.cpp
#include "vcg_client.hpp"
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>
#include <vector>

namespace vcg {

static int connect_dgram(const std::string& path, std::string& err) {
  int fd = ::socket(AF_UNIX, SOCK_DGRAM, 0);
  if (fd < 0) { err = "socket() failed"; return -1; }
  sockaddr_un addr{};
  addr.sun_family = AF_UNIX;
  std::snprintf(addr.sun_path, sizeof(addr.sun_path), "%s", path.c_str());
  if (::connect(fd, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) < 0) {
    err = "connect() failed to " + path; ::close(fd); return -1;
  }
  return fd;
}

Result await_window(const std::string& sock_path) {
  std::string err;
  int fd = connect_dgram(sock_path, err);
  if (fd < 0) return {false, err};
  std::vector<char> buf(65535);
  for (;;) {
    ssize_t n = ::recv(fd, buf.data(), buf.size(), 0);
    if (n <= 0) continue;
    try {
      auto msg = nlohmann::json::parse(std::string(buf.data(), n));
      if (msg.contains("type") && msg["type"] == "WINDOW_TICK") {
        ::close(fd);
        return {true, ""};
      }
    } catch (...) {}
  }
}

Result send_route(const std::string& from_d,
                  const std::string& to_d,
                  const std::string& payload,
                  const std::string& route_sock) {
  std::string err;
  int fd = connect_dgram(route_sock, err);
  if (fd < 0) return {false, err};
  nlohmann::json msg = {
    {"type","VCG_ROUTE"},
    {"from_d",from_d},
    {"to_d",to_d},
    {"payload",payload}
  };
  auto data = msg.dump();
  ::send(fd, data.data(), data.size(), 0);
  ::close(fd);
  return {true, ""};
}

std::optional<nlohmann::json> recv_route(const std::string& dimension_id) {
  std::string sock = "/run/vcg/" + std::string{dimension_id};
  for (auto& c : sock) c = std::tolower(c);
  sock += ".sock";
  std::string err;
  int fd = connect_dgram(sock, err);
  if (fd < 0) return std::nullopt;
  std::vector<char> buf(65535);
  ssize_t n = ::recv(fd, buf.data(), buf.size(), 0);
  ::close(fd);
  if (n <= 0) return std::nullopt;
  try {
    auto msg = nlohmann::json::parse(std::string(buf.data(), n));
    if (msg.contains("type") && msg["type"] == "VCG_ROUTE" && msg.contains("payload")) {
      return nlohmann::json::parse(msg["payload"].get<std::string>());
    }
  } catch (...) {}
  return std::nullopt;
}

} // namespace vcg

File: docs/VCG/src/gen1/vcg/clients/rust/vcg_client.m
# vcg_client – Rust helper for Gen1 workloads

## Purpose
Provide a minimal, reusable Rust API for communicating with the VCG orchestrator and Resonant Time Daemon.

---

## Functions

### `await_window(sock_path: &str) -> std::io::Result<()>`
Blocks until a `WINDOW_TICK` message is received for the current dimension.

### `send_route(from_d: &str, to_d: &str, payload: &str) -> std::io::Result<()>`
Sends a payload to another dimension via the VCG orchestrator.

### `recv_route(dimension_id: &str) -> std::io::Result<Option<serde_json::Value>>`
Receives the next routed payload for this dimension, if available.

---

## Implementation

```rust
use std::os::unix::net::UnixDatagram;
use std::path::Path;
use serde_json::{Value, json};
use std::fs;

const BUFFER_SIZE: usize = 65535;
const VCG_ROUTE_SOCK: &str = "/run/vcg/vcg_route.sock";

pub fn await_window(sock_path: &str) -> std::io::Result<()> {
    let sock = UnixDatagram::unbound()?;
    sock.connect(sock_path)?;
    let mut buf = vec![0u8; BUFFER_SIZE];
    loop {
        let size = sock.recv(&mut buf)?;
        let msg: Value = serde_json::from_slice(&buf[..size])?;
        if msg.get("type") == Some(&Value::String("WINDOW_TICK".into())) {
            return Ok(());
        }
    }
}

pub fn send_route(from_d: &str, to_d: &str, payload: &str) -> std::io::Result<()> {
    let sock = UnixDatagram::unbound()?;
    sock.connect(VCG_ROUTE_SOCK)?;
    let msg = json!({
        "type": "VCG_ROUTE",
        "from_d": from_d,
        "to_d": to_d,
        "payload": payload
    });
    let data = serde_json::to_vec(&msg)?;
    sock.send(&data)?;
    Ok(())
}

pub fn recv_route(dimension_id: &str) -> std::io::Result<Option<Value>> {
    let sock_path = format!("/run/vcg/{}.sock", dimension_id.to_lowercase());
    if !Path::new(&sock_path).exists() {
        return Err(std::io::Error::new(std::io::ErrorKind::NotFound, "Socket not found"));
    }
    let sock = UnixDatagram::unbound()?;
    sock.connect(&sock_path)?;
    let mut buf = vec![0u8; BUFFER_SIZE];
    let size = sock.recv(&mut buf)?;
    let msg: Value = serde_json::from_slice(&buf[..size])?;
    if msg.get("type") == Some(&Value::String("VCG_ROUTE".into())) {
        if let Some(payload) = msg.get("payload") {
            let parsed: Value = serde_json::from_str(payload.as_str().unwrap_or(""))?;
            return Ok(Some(parsed));
        }
    }
    Ok(None)
}
```
## Usage Example
```rust
use vcg_client::{await_window, send_route, recv_route};

fn main() -> std::io::Result<()> {
    let dim_id = "D2";
    let sock_path = format!("/run/vcg/{}.sock", dim_id.to_lowercase());

    await_window(&sock_path)?;
    send_route(dim_id, "D4", r#"{"msg": "hello"}"#)?;
    if let Some(msg) = recv_route(dim_id)? {
        println!("Received: {:?}", msg);
    }
    Ok(())
}
```
